static/js/dropdown_list_widget.js

```diff
@@ -238,3 +240,314 @@ DropdownListWidget.prototype.value = function () {
+
+// Remove the check mark from dropdown element.
+MultiSelectDropdownListWidget.prototype.remove_check_mark = function (element) {
+    const icon = element.find("i").expectOne();
+    const value = element.attr("data-value");
+    const index = this.data_selected.indexOf(value);
+
+    if (index > -1) {
+        icon.remove();
+        element.removeClass("checked");
+        this.data_selected.splice(index, 1);
+    }
+};
+
+MultiSelectDropdownListWidget.prototype.dropdown_focus_events = function () {
+    // Main keydown event handler which transfers the focus from one child element
+    // to another.
+
+    const search_input = $(`#${CSS.escape(this.container_id)} .dropdown-search > input[type=text]`);
+    const dropdown_menu = $(`.${CSS.escape(this.widget_name)}_setting .dropdown-menu`);
+    const filter_button = $(`#${CSS.escape(this.container_id)} .multiselect_btn`);
+
+    const dropdown_elements = () => {
+        const dropdown_list_body = $(
+            `#${CSS.escape(this.container_id)} .dropdown-list-body`,
+        ).expectOne();
+
+        return dropdown_list_body.children().find("a");
+    };
+
+    dropdown_menu.on("keydown", (e) => {
+        function trigger_element_focus(element) {
+            e.preventDefault();
+            e.stopPropagation();
+            element.trigger("focus");
+        }
+
+        switch (e.key) {
+            case "ArrowDown": {
+                switch (e.target) {
+                    case dropdown_elements().last()[0]:
+                        trigger_element_focus(filter_button);
+                        break;
+                    case $(`#${CSS.escape(this.container_id)} .multiselect_btn`)[0]:
+                        trigger_element_focus(search_input);
+                        break;
+                    case search_input[0]:
+                        trigger_element_focus(dropdown_elements().first());
+                        break;
+                }
+
+                break;
+            }
+            case "ArrowUp": {
+                switch (e.target) {
+                    case dropdown_elements().first()[0]:
+                        trigger_element_focus(search_input);
+                        break;
+                    case search_input[0]:
+                        trigger_element_focus(filter_button);
+                        break;
+                    case $(`#${CSS.escape(this.container_id)} .multiselect_btn`)[0]:
+                        trigger_element_focus(dropdown_elements().last());
+                        break;
+                }
+
+                break;
+            }
+            case "Tab": {
+                switch (e.target) {
+                    case search_input[0]:
+                        trigger_element_focus(dropdown_elements().first());
+                        break;
+                    case filter_button[0]:
+                        trigger_element_focus(search_input);
+                        break;
+                }
+
+                break;
```

Comment:
This function is very similar between the single-select and multi-select versions.  I'm OK with merging as-is, but it'd be a lot better to deduplicate them.  I think structurally it's a question of what our plans are for the "Filter" button at the bottom, which I'm not sure we actually need.

Reply:
As I mentioned earlier, since this widget will be implemented in recent-topic view (where I consider the keyboard focus event to be most important).
The focus event binded with filter button works in the following way as of now - 

* When triggered the ArrowUp key, the focus shifts to the last dropdown item.
* Upon pressing the ArrowDown/Tab key, the focus jumps to the seach input bar.

Reply:
But yes, it will be a lot better to deduplicate them. I couldn't come up with a cleaner way as of now which also considers the filter button focus events in case of multiselect dropdown.

Reply:
Yeah, we'd need to somehow pass a `next_element_above` into the initialization of the widget.